File: examples/embedded-c++/imbridge_test/tpcx_ai/uc06.cpp

#include "duckdb.hpp"

#include <chrono>
#include <iostream>
#include <string>
#include <thread>

using namespace duckdb;
using namespace imbridge;

static void udf_tmp(DataChunk &input, ExpressionState &state, Vector &result) {
	using TYPE = double;
	result.SetVectorType(VectorType::FLAT_VECTOR);
	auto result_data = ConstantVector::GetData<TYPE>(result);
	input.Flatten();
	auto tmp_data1 = ConstantVector::GetData<TYPE>(input.data[0]);
	auto tmp_data2 = ConstantVector::GetData<TYPE>(input.data[1]);
	memset(result_data, std::numeric_limits<int64_t>::min(), input.size() * sizeof(int64_t));
	for (idx_t i = 0; i < input.size(); i++) {
		result_data[i] = (int)tmp_data1[i];
	}
}

int main() {
	DuckDB db("/root/workspace/duckdb/examples/embedded-c++/imbridge_test/db/db_tpcx_ai_sf10.db");
	Connection con(db);
	con.CreateVectorizedFunction<int64_t, double, double, double, double, double, double, double>(
	    "udf", &udf_tmp, LogicalType::INVALID, FunctionKind::PREDICTION, 4096);

	string sql = R"(
explain analyze select serial_number, udf(smart_5_raw, smart_10_raw, smart_184_raw, smart_187_raw, smart_188_raw, smart_197_raw, smart_198_raw) 
from Failures;
)";
	int times = 5;
	double result = 0;
	double min1, max1;
	bool flag = true;
	for (int i = 0; i < times; i++) {
		printf("step times %d \n", i);
		clock_t start_time = clock();
		con.Query(sql);
		clock_t end_time = clock();
		double t = (double)(end_time - start_time) / CLOCKS_PER_SEC;
		printf("t%d :  %lf s!\n", i, t);
		result += t;
		if (flag) {
			flag = false;
			min1 = t;
			max1 = t;
		} else {
			min1 = std::min(min1, t);
			max1 = std::max(max1, t);
		}
	}
	result = result - min1 - max1;
	times = times - 2;
	printf("finished execute %lf s!\n", result / (times * 1.0));
	return 0;
}

File: examples/embedded-c++/imbridge_test/tpcx_ai/uc06.py
import duckdb
import joblib
import numpy as np
import pandas as pd
from duckdb.typing import BIGINT, DOUBLE, FLOAT, VARCHAR
from imblearn.over_sampling import ADASYN
from imblearn.pipeline import Pipeline as Imb_Pipeline
from sklearn import metrics
from sklearn.metrics import classification_report, confusion_matrix, f1_score
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from tqdm import tqdm
import time
# hand_type = "udf"
hand_type = "special"
name = "uc06"
scale = 10


con = duckdb.connect(
    f"/root/workspace/duckdb/examples/embedded-c++/imbridge_test/db/db_tpcx_ai_sf{scale}.db")

root_model_path = f"/root/workspace/duckdb/examples/embedded-c++/imbridge_test/data/tpcxai_datasets/sf{scale}"


model_file_name = f"{root_model_path}/model/{name}/{name}.python.model"


model = joblib.load(model_file_name)


def udf(smart_5_raw,
        smart_10_raw,
        smart_184_raw,
        smart_187_raw,
        smart_188_raw,
        smart_197_raw,
        smart_198_raw):
    data = pd.DataFrame({
        'smart_5_raw': smart_5_raw,
        'smart_10_raw': smart_10_raw,
        'smart_184_raw': smart_184_raw,
        'smart_187_raw': smart_187_raw,
        'smart_188_raw': smart_188_raw,
        'smart_197_raw': smart_197_raw,
        'smart_198_raw': smart_198_raw
    })
    # print(data.shape)
    return model.predict(data)

con.create_function("udf", udf, [DOUBLE, DOUBLE, DOUBLE, DOUBLE, DOUBLE, DOUBLE, DOUBLE], BIGINT, type="arrow", null_handling=hand_type)


sql = '''
explain analyze select serial_number, udf(smart_5_raw, smart_10_raw, smart_184_raw, smart_187_raw, smart_188_raw, smart_197_raw, smart_198_raw) 
from Failures;
'''

times = 5
min1 = 0
max1 = 0
res = 0
flag = True
for i in tqdm(range(times)):
    s = time.time()
    con.sql(sql)
    e = time.time()
    t = e-s
    print(t)
    res = res + t
    if flag:
        flag = False
        min1 = t
        max1 = t
    else:
        min1 = t if min1 > t else min1
        max1 = t if max1 < t else max1

res = res - min1 - max1
times = times - 2
print(f"{name}, {res/times}s ")
